### ISSUE TITLE
.NET 8 Blazor - Support localization change without page reload

### ISSUE THREAD

**Issue by joostvz** [dotnet/runtime#98415]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I'm trying to create a Blazor WebAssembly app where users can switch the display language. This only seems to work after forcing a reload of the page. According to the conversation in dotnet/aspnetcore#50361, switching languages without a page reload is currently not supported in Webassembly. I have a web app where it is not acceptable to reload the page when the user switches the language. A page refresh gives a bad user experience. And our app also contains a form, which data must be preserved when switching to another language.

I found a workaround by using this package: https://github.com/ScarletKuro/Blazor.WebAssembly.DynamicCulture, which calls an internal WebAssembly API: `INTERNAL.loadSatelliteAssemblies`. Normally, Blazor (WebAssembly) calls this method to load the assemblies for the current active culture. The _Blazor.WebAssembly.DynamicCulture_ package calls this method to load the assemblies of all specified cultures. After loading all specified assemblies, it is possible to change the language without a page reload.

### Describe the solution you'd like

Is it possible to open up the `INTERNAL.loadSatelliteAssemblies` API, so that developers can specify the cultures they want to be loaded? By default it can still load only the current culture, but it would be nice if developers can override this behavior.

**Comment by dvoituron** [comment]
I think you could add this C# code to call the equivalent JS code.
You have an example [here](https://github.com/dotnet/runtime/blob/main/src/mono/wasm/testassets/WasmBasicTestApp/App/SatelliteAssembliesTest.cs).
 
```csharp
[JSImport("INTERNAL.loadSatelliteAssemblies")]
internal static partial Task LoadSatelliteAssemblies(string[] culturesToLoad);
```

**Comment by joostvz** [comment]
@dvoituron Thanks for your response. This is indeed what the _Blazor.WebAssembly.DynamicCulture_ is also doing, and this approach works.

My only concern is that this is using an internal Blazor API, with zero guarantees. Hence my feature request to open up this API in a decent way.

**Comment by lewing** [comment]
There isn't anything in the runtime that restricts which satellite assemblies can be loaded and if they are added to the correct items I believe the build will correctly pick them up but blazor has some custom logic here so @maraf can you take a look and give appropriate guidance.

**Comment by maraf** [comment]
Making a public C# API is on our radar, but so far you are the first one asking for it.

In the meantime there are several workarounds
a) Use JSImport as mentioned above (the drawback is that the API might change between major .NET versions)
b) Load all satellite assemblies `Blazor.start({ configureRuntime: runtime => runtime.withConfig({ loadAllSatelliteResources: true }) })` (wich also uses internal switch)
c) If you know which assemblies contain resources, you fetch them from C#
```c#
var http = new HttpClient();
var dllBytes = await http.GetByteArrayAsync("/_framework/en-US/WasmBasicTestApp.resources.wasm");
AssemblyLoadContext.Default.LoadFromStream(new MemoryStream(dllBytes));
```

**Comment by joostvz** [comment]
Thanks for the provided workarounds @maraf. I will give them a try.

My scenario is the following: I'm developing a web app for a kiosk in a store. Users can select their preferred language on the screen. A page reload would probably cause a blinking page. Also, users should be able to switch the language at any moment in the process, even when they're filling in data in a form. A page reload would require all form data to be temporarily stored somewhere.

So for my scenario a page reload is possible, but has negative impact on the user experience, and adds technical challenges.

**Comment by maraf** [comment]
> My scenario is the following

Thank you! It makes perfect sense for your scenario

**Comment by alexspirou** [comment]
> Making a public C# API is on our radar, but so far you are the first one asking for it.
> 
> In the meantime there are several workarounds a) Use JSImport as mentioned above (the drawback is that the API might change between major .NET versions) b) Load all satellite assemblies `Blazor.start({ configureRuntime: runtime => runtime.withConfig({ loadAllSatelliteResources: true }) })` (wich also uses internal switch) c) If you know which assemblies contain resources, you fetch them from C#
> 
> ```cs
> var http = new HttpClient();
> var dllBytes = await http.GetByteArrayAsync("/_framework/en-US/WasmBasicTestApp.resources.wasm");
> AssemblyLoadContext.Default.LoadFromStream(new MemoryStream(dllBytes));
> ```
> 
> I'm curious what is your motivation for switching between cultures without reload? In a typical scenario users do this once and then use the prefered culture.

Hello @maraf, could you please provide guidance on where to write Blazor.start({ configureRuntime: runtime => runtime.withConfig({ loadAllSatelliteResources: true }) }) in a Blazor application?

**Comment by joostvz** [comment]
@alexspirou I'm currently using this approach, and did the following:
1. In your App.razor file, add `autostart="false"` to the include statement for the blazor.web.js file:
  `<script src="_framework/blazor.web.js" autostart="false"></script>`
1. Create a new javascript file where you customize the `start` method, and link this file in your App.razor:
   ```
   Blazor.start({
     webAssembly: {
         configureRuntime: dotnet => {
             dotnet.withConfig({ loadAllSatelliteResources: true });
         }
     }
   });
   ```

**Comment by alexspirou** [comment]
`<script src="_framework/blazor.web.js" autostart="false"></script>`

You mean to link the js here instead of _framework/blazor.web.js? Or in a new line like this : 

`    <script src="test.js"></script>
`

**Comment by alexspirou** [comment]
Thanks for the response @joostvz.

var dllBytes = await http.GetByteArrayAsync("/_framework/en-US/WasmBasicTestApp.resources.wasm");

 My resources are in a class library under the folder named Resources. How did you get the Assemblies?

**Comment by horaciodiez** [comment]
Hi! 
First, Kudos @joostvz for bringing and solving the issue  (and of course @ScarletKuro ). You saved my day.
Do you know if this issue is solved in .Net 10? 
Thanks!

**Comment by ScarletKuro** [comment]
> Hi!
> First, Kudos [@joostvz](https://github.com/joostvz) for bringing and solving the issue (and of course [@ScarletKuro](https://github.com/ScarletKuro) ). You saved my day.
> Do you know if this issue is solved in .Net 10?
> Thanks!

No, sadly they still didn't add the support, tho the `INTERNAL.loadSatelliteAssemblies` is capable of doing it.

But as @joostvz mentioned, it's internal undocumented API with zero guarantee, and each new TFM release the Blazor.WebAssembly.DynamicCulture breaks.

**Comment by horaciodiez** [comment]
Ooops. 
It still works in .net 9, i will test it in .net 10. 
Thak you!

**Comment by horaciodiez** [comment]
It doesn´t  wotk in .Net 10. Auch.

**Comment by ScarletKuro** [comment]
> It doesn´t wotk in .Net 10. Auch.

There is better alternative:
```diff
// index.html
-<script src="_framework/blazor.webassembly.js"></script>
+<script src="_framework/blazor.webassembly.js" autostart="false"></script>
<script>
+	Blazor.start({ configureRuntime: runtime => runtime.withConfig({ loadAllSatelliteResources: true }) })
</script>
```
```diff
// Program.cs
var host = builder.Build();

-await host.SetMiddlewareCulturesAsync(); //Can completely remove the dependency

await host.RunAsync();
```

**Comment by horaciodiez** [comment]
Hi @ScarletKuro ! Thank you very much. This was unexpected.

It works perfectly (even containerized in Azure) if i keep await host.SetMiddlewareCulturesAsync();
If I remove the line it doesn´t.

Thanks! 

> > It doesn´t wotk in .Net 10. Auch.
> 
> There is better alternative:
> 
> // index.html
> -<script src="_framework/blazor.webassembly.js"></script>
> +<script src="_framework/blazor.webassembly.js" autostart="false"></script>
> <script>
> +	Blazor.start({ configureRuntime: runtime => runtime.withConfig({ loadAllSatelliteResources: true }) })
> </script>
> // Program.cs
> var host = builder.Build();
> 
> -await host.SetMiddlewareCulturesAsync(); //Can completely remove the dependency
> 
> await host.RunAsync();
